### ISSUE TITLE
Is the polyfill safe to use when feature detection fails?

### ISSUE THREAD

**Issue by foolip** [GoogleChromeLabs/text-fragments-polyfill#157]
The [README](https://github.com/GoogleChromeLabs/text-fragments-polyfill/blob/main/README.md#usage) recommend this usage:

```js
// Only load the polyfill in browsers that need it.
if (
  !('fragmentDirective' in document)
) {
  import('text-fragments.js');
}
```

Due to https://webkit.org/b/273466 the polyfill will be loaded in Safari even though scroll to text fragment is supported, and the text is highlighted. Is this a problem, or is it harmless?

**Comment by tomayac** [comment]
(Just clarifying that the polyfill is never "safe" to use, as it modifies the DOM structure by inserting `<mark>`: https://github.com/GoogleChromeLabs/text-fragments-polyfill/blob/53375fea08665bac009bb0aa01a030e065c3933d/src/text-fragment-utils.js#L451-L512
The actual question is: does having undetectable native support paired with this polyfill make it even less safe.)

**Comment by tomayac** [comment]
Yes, I was about to say, the Venn diagram between "supports CSS Custom Highlight" and "does not support Text Fragments" needs to be studied carefully. And if a CSS Custom Highlight has the same challenges as the current `<mark>` wrapping. I did not look into that.

**Comment by foolip** [comment]
I wanted to test [the demo](https://text-fragments-polyfill.glitch.me/#:~:text=This%20domain,examples&text=More%20information...) in Safari but it looks like the deployment is old and doesn't include https://github.com/GoogleChromeLabs/text-fragments-polyfill/commit/2635dffc84fd8d666f27d580cd2af0fd3193434c.

**Comment by tomayac** [comment]
The demo pulls in https://unpkg.com/text-fragments-polyfill@5.7.0/dist/text-fragments.js.

**Comment by tomayac** [comment]
![Screenshot 2024-04-30 at 11 46 38](https://github.com/GoogleChromeLabs/text-fragments-polyfill/assets/145676/57c9b04e-b7a0-4517-a3e5-8664d3833880)

**Comment by foolip** [comment]
Yes, but it's guarded by the wrong feature detection:

```js
      if (!("fragmentDirective" in Location.prototype)) {
        import("https://unpkg.com/text-fragments-polyfill");
      }
```

**Comment by tomayac** [comment]
Oh, sorry. I fixed the glitch on Glitch. I thought you said it was wrong in the library. PTAL.

**Comment by foolip** [comment]
I see that the script loads in Safari still, but it doesn't seem to modify the DOM. Should I see `<mark>` somewhere?

**Comment by foolip** [comment]
Is that "luck" something the polyfill could depend on, to do away with the need for feature detection by other means?

**Comment by tomayac** [comment]
To be clear, the "luck" is [spec-mentioned behavior](https://wicg.github.io/scroll-to-text-fragment/#determine-if-fragment-id-is-needed:~:text=Similarly%2C%20since%20a%20window%E2%80%99s%20Location%20object%20is%20a%20representation%20of%20the%20URL%20of%20the%20active%20document%2C%20all%20getters%20on%20it%20will%20show%20a%20fragment%2Ddirective%2Dstripped%20version%20of%20the%20URL.) (this is why I put it in scare quotes).

If we're fine with the fact that a polyfill will be loaded for nothing, that is, it will run, but determine that there's no work to do and terminate, then we can say we don't need feature detection. The purist in me is not fine with this, but others may have differing opinions of course.

**Comment by foolip** [comment]
I was thinking about something like `location.hash.includes(':~:')` as the condition for loading the polyfill? Or wouldn't that work due to the timing of the stripping?

**Comment by tomayac** [comment]
My hunch is that this wouldn't work (as the stripping might happen before script has had a chance to run) or at best could be racy, but @bokand can definitely answer this better.

**Comment by bokand** [comment]
> My hunch is that this wouldn't work (as the stripping might happen before script has had a chance to run)

I think you want the stripping to happen for this? Assuming the polyfill reads the `:~:` directive and performs highlighting where the UA doesn't support it - if the directive is stripped it means the UA performs highlighting (polyfill unneeded). If it isn't stripped - the polyfill is needed. Per-spec, the stripping should happen before any scripts on the document are running. Do I understand correctly?

IMHO, feature detection seems more useful for a page deciding whether to append a `:~:` directive to anchors since a non-supporting UA won't strip the directive and it could potentially interact badly with script on the destination page.

However, it just came to me that pages can do a feature detection using:

```js
location.hash = ':~:text=test'; 
return location.hash = '' ? return 'supported' : 'unsupported';
```

This works in both Chrome and Safari, but assumes the page isn't listening to or using the fragment for anything else.

**Comment by foolip** [comment]
> Per-spec, the stripping should happen before any scripts on the document are running.

Thanks the key part I wasn't sure about, and that's great. It means that `location.hash.includes(':~:')` as the condition for loading the polyfill should be OK.

**Comment by tomayac** [comment]
The feature detection needs to be made more robust in case the hash is set. I think this could work:

```js
(() => {
  const oldHash = location.hash;
  return (location.hash = location.hash
    ? `${location.hash}:~:text=a`
    : "#:~:text=a`") && location.hash === oldHash
    ? true
    : false;
})();
```

Not sure if this would break really old `#!` hashbang Ajax sites, but I fail to find one to test.

**Comment by bokand** [comment]
https://groups.google.com :)
